netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java

```diff
@@ -86,36 +90,46 @@ public static ProtocolNegotiator tls(final SslContext sslContext,
     Preconditions.checkNotNull(sslContext, "sslContext");
     Preconditions.checkNotNull(inetAddress, "inetAddress");
 
-    final ChannelHandler sslBootstrapHandler = new ChannelHandlerAdapter() {
-      @Override
-      public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
-        // TODO(nmittler): This method is currently unsupported for OpenSSL. Need to fix in Netty.
-        SSLEngine sslEngine = sslContext.newEngine(ctx.alloc(),
-            inetAddress.getHostName(), inetAddress.getPort());
-        SSLParameters sslParams = new SSLParameters();
-        sslParams.setEndpointIdentificationAlgorithm("HTTPS");
-        sslEngine.setSSLParameters(sslParams);
-
-        SslHandler sslHandler = new SslHandler(sslEngine, false);
-        sslHandler.handshakeFuture().addListener(
-            new GenericFutureListener<Future<? super Channel>>() {
-              @Override
-              public void operationComplete(Future<? super Channel> future) throws Exception {
-                // If an error occurred during the handshake, throw it to the pipeline.
-                future.get();
-              }
-            });
-        ctx.pipeline().replace(this, "sslHandler", sslHandler);
-      }
-    };
+    final SslBootstrapHandler sslBootstrap = new SslBootstrapHandler(sslContext, inetAddress);
     return new ProtocolNegotiator() {
       @Override
       public Handler newHandler(Http2ConnectionHandler handler) {
-        return new BufferUntilTlsNegotiatedHandler(sslBootstrapHandler, handler);
+        return new BufferUntilTlsNegotiatedHandler(sslBootstrap, handler);
       }
     };
   }
 
+  @Sharable
+  private static class SslBootstrapHandler extends ChannelHandlerAdapter {
+    private final SslContext sslContext;
+    private final InetSocketAddress address;
+
+    SslBootstrapHandler(SslContext sslContext, InetSocketAddress address) {
+      this.sslContext = sslContext;
+      this.address = address;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+      // TODO(nmittler): This method is currently unsupported for OpenSSL in Netty < 4.1.Beta6.
+      SSLEngine sslEngine = sslContext.newEngine(ctx.alloc(),
```

Comment:
Would it be better to create the SslHandler within newHandler()? It seems all this code could be within newHandler() and we pass an sslHandler to BufferUntilTlsNegotiatedHandler instead of SslBootstrapHandler.